contrib/toolkits/XView/doc/man/cmdtool.1
.\" @(#)cmdtool.1 1.24 89/07/31 SMI; 
.TH CMDTOOL 1  "1 September 1989"
.SH NAME
.B cmdtool 
\- run a shell (or program) using the SunView text facility
.SH SYNOPSIS
.B cmdtool
[
.B \-C
]
[
.B \-M
.I bytes
]
[
.B \-P
.I count
]
[
.B \-B
.I boldstyle
]
[
.B \-I
.I command
]
[
.I generic-tool-arguments
]
[
.I program
[ 
.I program-arguments
]
]
.SH AVAILABILITY
Although this command is available with both the
SunView software installation option
and the
OpenWindows user environment,
its appearance is not the
same for both environments.
The man page you are currently
viewing refers to the
.B cmdtool 
that is available with OpenWindows. The primary
difference between this and the 
SunView 
.B cmdtool
is the user interface.  All OpenWindows standard tools
use the OPEN LOOK Graphical User Interface. For more
information about the OPEN LOOK UI, refer to the
.I OpenWindows User's Guide.
For information about installing OpenWindows, refer to the
.I OpenWindows Installation and Start-Up Guide.
.SH DESCRIPTION
.IX cmdtool "" "\fLcmdtool\fR \(em shell or program with SunView text facility"
.IX "shell window" "\fLcmdtool\fR"
.LP
.B cmdtool
is the standard
.I OpenWindows
support facility for shells and other programs.
When invoked,
.B cmdtool
runs a program (usually a shell) in a text-based command window.
Characters typed on the keyboard are inserted at the caret.  If the
program is a shell, that shell accepts and runs commands in the
usual manner.  
.B cmdtool
also supports programs that perform cursor motions directly, such
as
.BR vi (1).
.LP
The text of the current command line can be edited using normal
.BR textedit (1)
functions.
The command window displays a log of the
session, which can be scrolled through using the scrollbar
(unless the command does cursor motion).
This log can be edited, and saved by choosing the
.RB ` "Store as New File" '
item in the text facility's pop-up menu.
.SH OPTIONS
.TP 15
.B \-C
Console 
.BR cmdtool .
Redirect system console output to this shelltool.
Display console messages in this
.BR cmdtool ,
which might otherwise appear in unexpected places on the 
workstation screen.  Since a
.B cmdtool
window can be scrolled, console error messages can be recorded for
later examination.
.TP
.BI \-M " bytes"
Set the log to wrap-around after the indicated
number of
.IR bytes .
.TP
.BI \-P " count"
Checkpoint the log after every set of
.I count
editing operations.
.TP
.BI \-B " boldstyle"
Set the style for displaying bold text to
.IR boldstyle .
.I boldstyle
can be a string specifying one of the choices for the
.B /Tty/Bold_style
default, see
.BR "Defaults Options" ,
below, or it may be a numerical value for one of those choices,
from 0 to 8, corresponding to the placement of the choice in the list.
.TP
.BI \-I " command"
Pass
.I command
to the shell. 
.SM SPACE
characters within the command must be escaped.

.I generic-tool-arguments
.B cmdtool
accepts the generic tool arguments
listed in
.BR sunview (1).
.HP
.I program
[
.I program-arguments
]
.br
If a
.I program
argument is present,
.B cmdtool
runs it and passes any remaining arguments to that
.IR program .
If no
.I program
is given,
.B cmdtool
runs the program indicated by the
.SB SHELL
environment variable, or
.B /usr/bin/sh
by default.
.SH USAGE
.SS Defaults Options
.LP
The following options can be configured as default settings
using
.BR defaultsedit (1).
.TP
.B Term.EnableEdit:
When set to
.SB TRUE
(the standard default) only command lines can be edited.
.SB FALSE
allows the entire log to be edited.
(See also the description of 
.B Enable Edit
below.)
.TP
.B Text.InsertMakesCaretVisible:
This entry allows you to specify the method for displaying the
editing caret.
.RS
.PD 0
.TP 18
.B Same_as_for_text
Use the setting specified in the defaults for the 
.B Text
category (the standard default).
.TP
.B If_auto_scroll
If the caret is showing, and an inserted
.SM NEWLINE
would position it below the bottom of the screen 
(as determined by 
.BR /Text/Lower_context ),
the text is scrolled to keep the caret showing.
The number of lines scrolled is determined by the
.BR /Text/Auto_scroll_by
default.
(See
.BR textedit (1)
for more information.)
.br
.ne 3
.TP
.B Always
Scroll the caret back into view whenever input would position it
off the screen.
.PD
.RE
.TP
.B Text.CheckpointFrequency:
If set to
.B 0
(the standard default) no checkpointing is done.
For any value greater than zero,
a checkpoint is made each time the indicated number
editing operations has been performed since the last checkpoint.
Each character typed, each
.BR Paste ,
and each
.B Cut
counts as an editing operation.
At each checkpoint, an updated copy of the log is saved in a
file with a name that is constructed by appending two percent signs
.RB ( %%  )
to the name of the log file. 
By default, the log file has a name of the form
.BI /tmp/tty.txt. pid
.RI ( pid
is the process
.SM ID
number of
.BR cmdtool );
the corresponding checkpoint file has a name of the form
.BI /tmp/tty.txt. nnnnnn %% .
.TP
.B Text.MaxLogFileSize:
If set to
.B 0
(the standard default) no wrap-around takes place; the log file
grows without a specified limit. 
For values greater than zero,
wrap-around occurs whenever the indicated number of characters have
been written to the log since the last wrap-around.
Characters that are pushed over the top are replaced by the
message:
.RS
.IP
.B 
*** Text is lost because the maximum edit log size has been exceeded. ***
.RE
.TP
.B Keyboard.DeleteChar:
Set the character for erasing to the left of the caret.
Note:
in
.BR cmdtool ,
the
.RB ` "stty erase" '
command has no effect.
Text-based tools refer only to the defaults database key settings.
The default is 
.SM DELETE\s0.
.TP
.B Keyboard.DeleteWord:
Set the character for erasing the word to the left of the caret.
The standard default is 
.SM CTRL-W\s0.
.TP
.B Keyboard.DeletetoLine:
Set the character for erasing all characters to the left of
the caret.  Note:
.RB ` "stty kill" '
has no effect in
.BR cmdtool .
The standard default is 
.SM CTRL-U\s0.
.LP
.B cmdtool Windows
.LP
The window created by 
.B cmdtool
is based on the text facility that is described in the 
.B textedit
man page.
The user is given a prompt at which to type commands and pop-up
menus from which to select command options.
.LP
.B cmdtool
windows support cursor motions, using an
.B /etc/termcap
entry called
.BR sun-cmd .
Command windows automatically set the
.SB TERM
environment variable to
.BR sun-cmd .
So, if you
.BR rlogin (1C)
to a machine that does not have an entry for
.B sun-cmd
in its
.B /etc/termcap
file, the error message
.RB ` "Type sun-cmd unknown" '
results.
To rectify this, type the command
.RB ` "set \s-1TERM\s0=sun" '.
Programs written using the
.BR curses (3X)
or 
.BR curses (3V)
library packages will work in a command window, but programs
hard-coded for
.BR sun -type
terminals may not. 
When supporting a program that performs
cursor motions, the command window automatically takes on the
characteristics of a tty window (as with
.BR shelltool (1)).
When that program terminates or sleeps, the full command window
functionality is restored.
.LP
.B cmdtool
supports programs that use
.SM CBREAK
and
.SM NO ECHO
terminal modes. 
This support is normally invisible to the user.
However, programs that use
.SM RAW
mode, such as
.BR rlogin (1C)
and
.BR script (1),
inhibit command-line editing with the mouse.
In this case, however, tty-style
.SM ERASE,
word-kill and line-kill characters can still be used to edit the
current command line.
.SS The cmdtool Menu
.LP
The
.B cmdtool
window menu is called the
.B Term Pane
menu and contains the following options and their submenus:
.RS
.B History 
.RS
Creates a a list of commands used during the
.B cmdtool session
.RE
.B Edit >
.RS
Provides a set of editing functions as described below:

.B Again
.RS
Repeats the last action.
.RE
.B Undo >
.RS
.B Undo Last Edit
.RS
Undoes the last edit made in
.B cmdtool.
.RE
.B Undo All Edits
.RS
Undoes all edits made during this session of
.B cmdtool.
.RE
.RE
.B Copy
.RS
Makes a copy of the selected text and places it in the clipboard.
.RE
.B Paste
.RS
Pastes a copy of the text stored in the clipboard at the cursor location.
.RE
.B Cut
.RS
Deletes the selected text and places it in the clipboard.
.RE
.RE
.B Find >
.RS
Provides a set of find and replace functions as described below:

.B Find and Replace
.RS
Brings up a pop-up menu containing text fields and command
buttons that allow you to search forward and backward through
the file being edited for specific text strings. Allows you
to specify options for the replacement of text.
.RE
.B Find Selection >
.RS
.B Forward
.RS
Searches forward to find a selected text string.
.RE
.B Backward
.RS
Searches backward to find a selected text string.
.RE
.RE
.B Find Marked Text
.RS
Brings up a pop-up menu that allows you to find text that is
included between specified bracket pairs. Also allows you to
insert or remove bracket pairs around selected text.
.RE
.B Replace |>field<| >
.RS
Allows you to replace selected text forward and backward throughout
the file.
.RE
.RE
.B Extras >
.RS
A user-definable pull-right menu conrolled by the
.B /usr/lib/text_extras_menu
file.
.RE
.B File Editor >
.RS
.B Enable
.RS
Allows you to edit files from within
.B cmdtool.
.RE
.B Disable
.RS
Turns off the ability to edit files from
within 
.B cmdtool.
.RE
.RE
.B Scrolling >
.RS
Enables scrolling within
.B cmdtool.
.RE
.B Disable Scrolling >
.RS
Turns off the ability to scroll within
.B cmdtool.
Once scrolling in 
.B cmdtool is disabled, it's functionality is
identical to
.B shelltool
and a more restricted menu appears. Selecting the
.B Enable Scrolling
option from the restricted menu restores the full menu
and functionality of
.B cmdtool.
.RE
.RE

.SS Accelerators
.LP
Text facility accelerators that are especially useful in
command windows are described here.  See 
.br
.BR textedit (1)
for more information.
.TP 18
.SM CTRL-RETURN
Position the caret at the bottom, and scroll it into view
as determined by 
.BR Text.LowerContext .
.TP 
.PD 0
\s-1CAPS\s0-lock
.TP
Toggle between all-upper-case keyboard input, and mixed-case.
.PD
.SH FILES
.PD 0
.TP 20
.BI /tmp/tty.txt. pid
log file
.TP
.B ~/.textswrc
.TP
.B ~/.ttyswrc
.TP
.B usr/lib/.text_extras_menu
.TP
.B $HOME/.TextExtraMenu
.TP
.B /etc/termcap
.TP
.B /usr/bin/sh
.PD
.SH "SEE ALSO"
.BR defaultsedit (1),
.BR rlogin (1C),
.BR script (1),
.BR shelltool (1),
.BR sunview (1),
.BR textedit (1),
.BR vi (1),
.BR curses (3V),
.BR curses (3X)
.LP
.TX INSTALL
.br
.TX SVBG
.SH BUGS
.LP
Typing ahead while
.B cmdtool 
changes between its scrollable and cursor motion modes
will sometimes freeze
.B cmdtool .
.LP
Full terminal emulation is not complete.
Some manifestations of this deficiency are:
.TP 3
\(bu
File completion in the C shell does not work.
.TP
\(bu
Enhanced display of text is not supported.


contrib/toolkits/XView/clients/cmdtool/cmdtool.c
#ifndef lint
#ifdef sccs
static  char sccsid[] = "@(#)cmdtool.c 15.36 89/10/26";
#endif
#endif

/*
 * Copyright (c) 1985, 1987 by Sun Microsystems, Inc.
 */

/*
 *  cmd/shelltool - run a process in a tty subwindow
 */

#include <stdio.h>
#include <strings.h>
#include <sys/types.h>
#include <xview/attr.h>
#include <xview/defaults.h>
#include <xview/xview.h>
#include <xview/scrollbar.h>
#include <xview/tty.h>
#include <xview/termsw.h>
#include <xview/textsw.h>
#ifdef ecd.help
#include <xview/help.h>
#endif

#ifdef STANDALONE
#define EXIT(n)		exit(n)
#else
#define EXIT(n)		return(n)
#endif

#define HEIGHTADJUST 19

extern	char *getenv();

#ifdef SUNVIEW1
extern  int  ttysw_is_already_dead();
#endif

static short tty_image[258] = {
#include <images/terminal.icon>
};
mpr_static(tty_pixrect, 64, 64, 1, tty_image);

static short cmd_image[258] = {
#include <images/cmdtool.icon>
};
mpr_static(cmd_pixrect, 64, 64, 1, cmd_image);

print_usage(am_cmdtool, toolname)
	int	 am_cmdtool;
	char	*toolname;
{
	char	*mode_spec = (am_cmdtool)
				? "-P frequency] [-W size" :
				"-B boldstyle";
	
	(void)fprintf(stderr,
		"syntax: %s [-C] [-I initial_cmd] [%s] [program [args]]\n",
		toolname, mode_spec);
	if (!am_cmdtool) {
	    (void)fprintf(stderr,
		"-B	set bold style for this instance of %s\n", toolname);
	    (void)fprintf(stderr,
		"	where 'boldstyle' is a number from 1 to 8\n");
	}
	(void)fprintf(stderr,
		"-C	redirect console output to this instance of %s\n",
		toolname);
	(void)fprintf(stderr,
		"-I	'initial_cmd' is first command 'program' executes\n");
	if (am_cmdtool) {
	    (void)fprintf(stderr,
		"-P	checkpoint frequency for this %s, %s\n",
		toolname, "where 'frequency' is number");
	    (void)fprintf(stderr,
		"	of edits between checkpoints; %s\n",
		"a value of 0 means no checkpointing.");
	    (void)fprintf(stderr,
		"-W	wrap edit log for this %s %s\n", toolname,
		"at 'size' bytes; size is either");
	    (void)fprintf(stderr,
		"	at least 8096, or 0 which means no wrapping.\n");
	}
}

#ifdef STANDALONE
main(argc,argv)
#else
int shelltool_main(argc, argv)
#endif STANDALONE
	int argc;
	char **argv;
{
	int	am_cmdtool;
	Frame	base_frame;
	Tty	ttysw;
	Icon	tool_icon;
	char	*tool_name = argv[0];
	char	*shell_label = "shelltool";
	char	*cmd_label = "cmdtool";
	char	*console_label = " (CONSOLE) - ";
	char	frame_label[150];
	char	icon_label[30];
        char    *tmp_label1, *tmp_label2;
	int	become_console = 0;
	char	*bold_name = 0;
	char	*sh_argv[2];
	char	*init_cmd = 0;
	int	len;
	char	*filename = (char *)rindex(argv[0], '/');
	int	checkpoint = 0;
	int	edit_log_wraps_at = TEXTSW_INFINITY;
	int	tty_pid = 0;
	int	height_is_set, width_is_set;
	char	err_msg[50];
	static  Notify_value my_destroy_func();
	char    cmdline[50];
	
#ifdef GPROF
	if (argc > 1 && strcmp(argv[argc-1], "-gprof") == 0) {
	    moncontrol(1);
	    /* Pull the -gprof out of argc/v */
	    argc--;
	    argv[argc] = (char *)0;
	} else {
	    moncontrol(0);
	}
#endif
	
#ifdef DEBUG
	malloc_debug(0);   
#endif	
	
	/* This is required to initialize correctly */	
	xv_init(XV_INIT_ARGC_PTR_ARGV, &argc, argv, 0);

	if (filename)
		filename++;
	else
		filename = argv[0];
	am_cmdtool = (strcmp(filename, "cmdtool") == 0);

	width_is_set =  am_cmdtool && 
		((defaults_exists("cmdtool.columns", "window.columns")) ||
		(defaults_exists("cmdtool.width", "window.width")));
	height_is_set = (defaults_exists("cmdtool.rows", "window.rows")) ||
		(defaults_exists("cmdtool.height", "window.height"));

	/*
	 *  Send the icon attr before argc, argv to give
	 *  commandline argument a chance to override.
	 *  A waste of space & time if commandline argument
	 *  is present.
	 */
	icon_label[0] = 0177;	/* del, highly unlikely as cmd arg */
	icon_label[1] = '\0';
	tool_icon = xv_create((Xv_window)NULL, ICON,
			ICON_IMAGE,	am_cmdtool ?
					&cmd_pixrect : &tty_pixrect,
			ICON_LABEL,	icon_label,
			0);

	base_frame = xv_create((Xv_window)NULL, FRAME,
			FRAME_ICON,		tool_icon,
			0);
			
	if (base_frame == NULL) {
	    fprintf(stderr, "Cannot create base frame.  Process aborted.\n");
	    EXIT(1);
	}		

	/* Get ttysw related args */
	sh_argv[0] = NULL;
	sh_argv[1] = NULL;
	argv++;
	argc--;
	if (am_cmdtool) {
	    checkpoint =
		defaults_get_integer_check("cmdtool.checkpointFrequency",
			"Term.CheckpointFrequency", 0, 0, (int)TEXTSW_INFINITY);
	    edit_log_wraps_at =
		defaults_get_integer_check("cmdtool.maxLogFileSize",
			"Term.MaxLogFileSize",
			(int)TEXTSW_INFINITY, 0, (int)TEXTSW_INFINITY);
	}
	cmdline[0] = NULL;
	while (argc > 0 && **argv == '-') {
		switch (argv[0][1]) {
		case 'C':
			become_console = 1;
			strcat(cmdline, " -C ");
			break;
		case '?':
			(void)xv_usage(tool_name);
			print_usage(am_cmdtool, tool_name);
			EXIT(1);
		case 'B':
			if (argc > 1) {
				argv++;
				argc--;
				bold_name = *argv;
			}
			break;
		case 'I':
			if (argc > 1) {
				argv++;
				argc--;
				init_cmd = *argv;
			}

			break;
		case 'P':
			checkpoint = atoi(argv[1]);
			strcat(cmdline, " -P ");
			strcat(cmdline, argv[1]);
			argc--, argv++;
			break;
		case 'M':
			edit_log_wraps_at = atoi(argv[1]);
			strcat(cmdline, " -M ");
			strcat(cmdline,  argv[1]);
			argc--, argv++;
			break;
		default:
			;
		}
		argv++;
		argc--;
	}

	if (strlen(cmdline) > 0)
	    xv_set(base_frame, WIN_CMD_LINE, cmdline, 0);

	if (argc == 0) {
		argv = sh_argv;
		if ((argv[0] = getenv("SHELL")) == NULL)
			argv[0] = "/bin/sh";
	}

	/* If FRAME_LABEL wasn't set by cmdline argument, set it */
	if ((tmp_label1 = (char *)xv_get(base_frame, FRAME_LABEL)) == NULL) {
		(void)strncpy(frame_label,
		  am_cmdtool ? cmd_label : shell_label, sizeof(frame_label));
		if (become_console) {
			(void)strncat(frame_label, console_label,
				sizeof(frame_label));
		} else {
			(void)strncat(frame_label, " - ", sizeof(frame_label));
		}
		(void)strncat(frame_label, *argv, sizeof(frame_label));
		(void)xv_set(base_frame, FRAME_LABEL, frame_label, 0);
	}
	tool_icon = (Icon)xv_get(base_frame, FRAME_ICON);
	if (((tmp_label2 = (char *) xv_get(tool_icon, ICON_LABEL)) == NULL) ||
	    *tmp_label2 == 0177) {
	    if (tmp_label1) {
		(void)strncpy(icon_label, tmp_label1, sizeof(icon_label));
            } else if (become_console) {
		(void)strncpy(icon_label, "CONSOLE", sizeof(icon_label));
	    } else {
		(void)strncpy(icon_label, *argv, sizeof(icon_label));
	    }
	    (void)xv_set(tool_icon, ICON_LABEL, icon_label, 0);
	}
	(void)xv_set(base_frame,
		     FRAME_ICON,		tool_icon,
#ifdef ecd.help
		     HELP_DATA, (become_console ? "xview:console" :
				    (am_cmdtool ? "xview:cmdtool" :
						"xview:shelltool")),
#endif
		     0);

	
	ttysw = xv_create(base_frame, TERMSW,   WIN_IS_CLIENT_PANE,
		  TTY_ARGV,			argv,
		  TTY_QUIT_ON_CHILD_DEATH,	TRUE,
		  TTY_CONSOLE,			become_console,
		  0);

	if (!width_is_set) {
		xv_set(ttysw,
		       WIN_COLUMNS, 80,
		       0);
		window_fit_width(base_frame);
	}

	/* only correct the height if the user did not give a pixel value
 	 * for the height.
	 */
	if (!height_is_set) {
	    xv_set(ttysw,
		   WIN_ROWS, 35,
		   0);
	    window_fit_height(base_frame);
	}

        if (!am_cmdtool)
		xv_set(ttysw, TERMSW_MODE, TTYSW_MODE_TYPE, 0);
	if (bold_name) {
		(void)xv_set(ttysw, TTY_BOLDSTYLE_NAME, bold_name, 0);
	}
	if (am_cmdtool) {
	    (void) xv_set(ttysw,
			  TEXTSW_CHECKPOINT_FREQUENCY, checkpoint,
			  TEXTSW_WRAPAROUND_SIZE, edit_log_wraps_at,
			  0);
	}
	
	tty_pid = (int)xv_get(ttysw, TTY_PID);
#ifdef DEBUG
	(void)fprintf(stderr, "child pid = %d\n", tty_pid);
#endif DEBUG
	if (tty_pid == -1) {
	    strcpy(err_msg, (am_cmdtool) ? "Command" : "Shell");
	    strcpy(err_msg, " Tool: Out of swap space.  Cannot continue.\n");
	    (void) ttysw_output(ttysw, err_msg, strlen(err_msg));        
	} else if (init_cmd && ((len = strlen(init_cmd)) > 0)) {
	    if (init_cmd[len-1] != '\n') {
		init_cmd[len] = '\n';
		len++;
	    }
	    (void)ttysw_input(ttysw, init_cmd, len);
	}

	xv_main_loop(base_frame);
	
	EXIT(0);
}

